apps/ensindexer/src/api/lib/indexing-status.ts

```
@@ -0,0 +1,374 @@
+          );
+        }
+
+        const historicalCompletedBlocks =
+          metrics.getMetricValue("ponder_historical_completed_blocks", {
+            chain: chainName,
+          }) ?? 0;
+
+        const hasSyncBackfill = historicalTotalBlocks > 0;
+
+        // If the chain has a sync backfill but hasn't completed any blocks,
+        // the chain has not started yet.
+        if (hasSyncBackfill && historicalCompletedBlocks === 0) {
+          return {
+            chainId,
+            result: {
+              status: "not_started" as const,
+              config: chainBlockConfig,
+            } satisfies ChainIndexingStatus,
+          };
+        }
+
+        // The backfill end block is equal to the earliest start block
+        // plus the total number of blocks in the backfill.
+        const backfillEndBlockNumber = chainBlockConfig.startBlock.number + historicalTotalBlocks;
+
+        // This will be cached after the first request.
+        const backfillEndBlock = await cachedGetBlockByNumber(chainName, backfillEndBlockNumber);
+
+        // During the backfill, the latest known block is the backfill end block.
+        const latestKnownBlock = backfillEndBlock;
+
+        return {
+          chainId,
+          result: {
+            status: "backfill" as const,
+            config: chainBlockConfig,
+            latestIndexedBlock,
+            latestKnownBlock,
+            backfillEndBlock,
+          } satisfies ChainIndexingStatus,
+        };
+      }),
+    );
+
+    const result = {
+      chains: Object.fromEntries(chainResults.map(({ chainId, result }) => [chainId, result])),
+    } satisfies IndexingStatus;
+
+    return c.json(result);
+  } catch (error) {
+    return c.json({ error: "Internal server error" }, 500);
+  }
+});
+
+const blockCache = new Map<string, Map<number, BlockRef>>();
+
+async function cachedGetBlockByNumber(chainName: string, blockNumber: number) {
```

Strongly prefer we take a strategy where caching of these results is irrelevant.

Goal: The only RPC calls we should need for this API are to get the block timestamps for the `config` and this should only happen once upon process startup. If this fails upon startup it should throw an error and force the entire process to crash. We must take a strategy where it is guaranteed that zero RPC calls are in the codepath of requests to this API endpoint.